### PR TITLE
[native] Add first version of json function signature parser

### PR DESCRIPTION
**Pull request by pedroerp**
Summary:
Adding a first version of the code to parse json function signature
files. It will be used to read json registration files generated for remote
functions.

Differential Revision: D47201915

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/JsonSignatureParser.h [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+/// This class can be conveniently used in a range for loop:
+///
+///   for (const auto& it : JsonSignatureParser(jsonString)) {
+///     // registration code
+///   }
+class JsonSignatureParser {
+ public:
+  JsonSignatureParser(const std::string& input);
+
+	// Iterator helpers.
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/JsonSignatureParser.h:56]: Wrong indentation?

File: presto-native-execution/presto_cpp/main/tests/JsonSignatureParserTest.cpp [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+namespace facebook::presto::test {
+namespace {
+
+class JsonSignatureParserTest : public testing::Test {};
+
+TEST_F(JsonSignatureParserTest, broken) {
+  // Needs to parse and provide a top level `udfSignatureMap` object.
+  EXPECT_THROW(JsonSignatureParser parser(""), VeloxUserError);
+  EXPECT_THROW(JsonSignatureParser parser("{}"), VeloxUserError);
+  EXPECT_THROW(
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/tests/JsonSignatureParserTest.cpp:30]: Use VELOX_ASSERT_THROW

File: presto-native-execution/presto_cpp/main/JsonSignatureParser.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  size_t size() const {
+    return signaturesMap_.size();
+  }
+
+  auto begin() const {
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/JsonSignatureParser.h:61]: 'auto' is kind of difficult to understand in a public API. Would it be possible to specify the type explicitly?

File: presto-native-execution/presto_cpp/main/CMakeLists.txt [hunk 4]
```diff
@@ -87,6 +87,10 @@ target_link_libraries(
   velox_tpch_connector
   velox_presto_serializer)
 
+add_library(presto_json_signature_parser JsonSignatureParserTest.cpp)
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/CMakeLists.txt:90]: Shouldn't this be defined in tests/CMakeLists.txt?

File: presto-native-execution/presto_cpp/main/JsonSignatureParser.h [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  TContainer::const_iterator end() const {
+    return signaturesMap_.end();
+  }
+
+ private:
+  /// Parses the top level json parsed.
+  void parse(const folly::dynamic& input);
+
+  /// Parses a list of signatures for a function.
+  std::vector<velox::exec::FunctionSignaturePtr> parseSignatures(
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/JsonSignatureParser.h:78]: Do these needs to be methods or can they be free functions in the .cpp file?

File: presto-native-execution/presto_cpp/main/tests/JsonSignatureParserTest.cpp [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+  })V";
+
+  JsonSignatureParser parser(input);
+  EXPECT_EQ(1, parser.size());
+
+  const auto& it = parser.begin();
+  EXPECT_EQ(it->first, "my_func");
+  EXPECT_EQ(it->second.size(), 1);
+
+  // We can just verify the counts here. If the type doesn't not exist or can't
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/tests/JsonSignatureParserTest.cpp:136]: typo: "doesn't not exit"

**Review by mbasmanova — APPROVED** [review]
@pedroerp Looks good to me. It would be nice to mentions unsupported features, e.g. generic signatures, constant arguments, variadic arguments, aggregate functions, etc.

**Comment by pedroerp** [comment]
Merged in #20323

**Pull request closed without merging**